Cypress tests

sergesoroka opened the pull request:
Added Cypress tests: both the e2e and the Components ones to integrate into CI/CD pipeline.

kerberizer commented on cypress/e2e/basic.cy.js:
```diff
@@ -1,19 +1,51 @@
+      {
+        fixture: "json/bk_rcapi_template_default.json",
+      }
+    ).as("getAllTemplates");
```
Why do we need this when we still define the intercepts in the tests themselves?

kerberizer commented on cypress/e2e/basic.cy.js:
```diff
@@ -1,19 +1,51 @@
+
+    cy.intercept("GET", "/template/*", {
+      fixture: "json/bk_rcapi_template_default.json",
+    }).as("getTemplate");
```
Shouldn't this define the hostname as well?

kerberizer commented on cypress/e2e/basic.cy.js:
```diff
@@ -1,19 +1,51 @@
+
+    cy.wait("@getTemplate")
+      .its("response.statusCode")
+      .should("be.oneOf", [200, 304]);
```
I'm not sure I understand this correctly. Since we're using a fixture here, why do we wait for the request and why do we check the status codes—this is something that Cypress itself will return as a canned response, so we have full control over what and how it will return. Sorry if I'm missing something here!

kerberizer commented on cypress/e2e/basic.cy.js:
```diff
@@ -1,19 +1,51 @@
+
+    cy.intercept("POST", "/template/*", {
+      fixture: "json/bk_rcapi_template_default.json",
+    }).as("saveTemplate");
```
Same as above.

kerberizer commented on cypress/e2e/basic.cy.js:
```diff
@@ -1,19 +1,51 @@
+      expect(request.body).to.include("template_name");
+      expect(request.body).to.include("template_author");
+      expect(request.body).to.include("template_acknowledgment");
+    });
```
Same as above. I'm under the impression—which could be wrong—that we're testing what the backend will return. But we aren't testing the backend. That's why we want fixtures, so the backend could be completely eliminated from the tests, emulating as if it always returns correct (and expected) responses. This allows us to test only the functionality of the frontend we're developing here. Again, I apologize if I'm stupidly missing something obvious.

sergesoroka replied:
You're right. I'll remove this checking.

kerberizer commented on package.json:
```diff
@@ -36,6 +36,7 @@
     "zustand": "^4.5.0"
   },
   "devDependencies": {
+    "@types/cypress": "^1.1.6",
```
Do we really need this? https://www.npmjs.com/package/@types/cypress says it's deprecated and included in cypress itself.

kerberizer commented on src/App.cy.jsx:
```diff
@@ -0,0 +1,27 @@
+    //   expect(interception.response.statusCode).to.eq(200);
+    // });
+  });
+});
```
I guess this was added here by the Cypress' "Create from component" tool. I don't know why Cypress does this, but tests definitely not be added to our application source. The proper location, I believe, is `cypress/component`.

I have the feeling that we're not comfortable yet with how component testing works, so maybe we should skip it for now? It would help debugging problems, but our priority right now is the user experience, which E2E tests.

May I suggest we split component testing in possibly a separate branch and pull request?

sergesoroka replied:
May I am mistaken but Nina insisted of using Component tests as well. She mentioned TDD (Test Driven Development), so I started writing these tests. As I saw it earlier in React app it common place to put a test beside a component itself, but of course it would be seperate directory e. g. `tests`. And we can use separate branch, too.

kerberizer replied:
Regarding the test files location, apparently there's a concept of [collocation](https://povio.com/blog/maintainability-with-colocation), which puts unit tests next to the components they test. I'll need to read more about it to form an opinion. It might actually be a good opportunity to review our directory structure (also for spectrasearch perhaps).

As for TDD, I believe @vedina primarily meant E2E tests. The rationale is that this is mostly a user interface application that does—for now at least—minimal data processing. Our main workflows that we need to test—and to design our application around—are these user interactions. It would be different for a library where the implemented algorithms would be the main functionality and E2E testing might not even make sense.

Of course, this doesn't mean that component testing is not desired. My point is that it is only secondary and mostly helpful to ourselves: proper component testing might help us determine why an E2E test fails (and, on rare occasions, might even catch some problems that weren't evident from the existing E2E tests).

To summarize, I still think it is best to separate the component tests in a different branch. It would be better even just for the review process.

sergesoroka replied:
Ok, let's focus on E2E tests for now.

kerberizer commented on cypress/fixtures/json/bk_rcapi_template_default.json:
```diff
@@ -1,28 +1,28 @@
+      "template_author": "Sergey",
+      "template_acknowledgment": "Test"
+    }
+  ]
```
Isn't it better to keep a real life example here? I think in the future we might also want to add different examples representing different edge cases. But I'm not sure I understand what is the advantage of this change.

kerberizer replied:
@sergesoroka Let's keep only entries that were created by us, e.g., Nina, Geri, Nick, etc. If two entries are enough for testing, we can use the file I've put there. If more that two entries are needed, let's add something entered by our people. I'm worried about having other people's data in our tests.

kerberizer commented on tsconfig.json:
```diff
@@ -4,6 +4,7 @@
       "@/*": ["./src/*"]
-    }
+    },
+    "esModuleInterop": true
```
Why is this needed?

kerberizer replied:
@sergesoroka Could you please see this?

kerberizer reviewed (changes requested):
@sergesoroka thank you very much for working on this. I haven't gone through all the files, because some of my comments concern the overall direction and we might want to completely rework the pull request. So let's start the discussion.

sergesoroka commented on cypress/e2e/basic.cy.js:
```diff
@@ -1,19 +1,51 @@
-      fixture: 'json/bk_rcapi_template_default.json',
-    })
-    cy.visit(testURLRoot)
+const testURLRoot = "http://localhost:5173/templates";
```
It seems to be very important. Thank you for explaining me.

sergesoroka commented on src/ui/Button.cy.tsx:
```diff
@@ -0,0 +1,10 @@
+    cy.mount(<Button label="View" disabled={false} />);
+    cy.get("[data-cy=View]").should("have.text", "View");
+  });
+});
```
removed

sergesoroka replied:
removed as well

sergesoroka commented on src/ui/Select.cy.tsx:
```diff
@@ -0,0 +1,11 @@
+    // @ts-ignore
+    cy.mount(<Select url={url} projectName="" />);
+  });
+});
```
removed as well

sergesoroka replied:
removed

kerberizer commented on cypress/e2e/general.cy.js:
```diff
@@ -0,0 +1,97 @@
+  it("can open an existing draft for editing", () => {
+    cy.get('[data-cy="draft"]').click();
+    cy.get(".nonSelected td").eq(1).click();
+    cy.get('[data-cy="Edit blueprint"]').click();
```
@sergesoroka This tests seems to be destined to fail without a proper intercept that would load the existing draft data from a fixture rather than from the non existing backend. I'm not sure how it seemed to work before the test files reorganization.

TL;DR: To properly test this, I believe we should have a proper intercept/fixture for the specific GET to `/template/{UUID}`.

kerberizer commented on cypress/e2e/general.cy.js:
```diff
@@ -0,0 +1,97 @@
+    cy.get(".nonSelected td").eq(1).click();
+    cy.get('[data-cy="Edit blueprint"]').click();
+
+    cy.get("#Save").click();
```
@sergesoroka Why do we need to click "Save" here? It's not a bad idea to test saving, but to be meaningful, we should intercept the POST request that we'll make and check its payload for the expected JSON.

kerberizer commented on cypress/e2e/general.cy.js:
```diff
@@ -0,0 +1,97 @@
+  it("can find an existing draft by searching", () => {
+    cy.get(".search").click().type("hts_metadata_test_finalized");
+    cy.get(".nonSelected td").eq(1).click();
+  });
```
@sergesoroka Are we certain here that we're getting the expected search result?

kerberizer replied:
I'm not happy with this test, but let's keep it this way for now and improve it after we merge this PR.

kerberizer commented on cypress/e2e/general.cy.js:
```diff
@@ -0,0 +1,97 @@
+    cy.get('[data-cy="current-page-number"]').should("have.text", "1");
+  });
+
+  it("can sort by a table column", () => {
+    cy.get('[data-cy="draft"]').click();
+    cy.get('[data-cy="Acknowledgment"]').click();
+    cy.get('[data-cy="Acknowledgment"]').click();
+    cy.get('[data-cy="Acknowledgment"]').click();
+  });
```
@sergesoroka Being able to click the column header is a good start, but we definitely want to check if the sorting actually produces the expected result too.

kerberizer replied:
Please scrap this test here. You've already created a separate test for sorting.

kerberizer reviewed (changes requested):
@sergesoroka I think we're gradually getting in good shape with this, though it also becomes evident that we still have a lot of work even with E2E testing alone.

I've taken the liberty to reorganize the tests. Let's keep them for now in a single file.

We'll branch individual test files from it once the number of tests becomes sufficiently large. For example, once/if we have several tests for sorting, we'll put them in a separate file. Let's agree on having at least 3 tests in a specific area (e.g., sorting, searching, editing) before we put them in a separate file. The rest of the tests will stay in `general`.

Here are a few important points in my view:
- Let's focus on completing the existing tests before adding new ones.
- Make sure that each test is indeed complete. For example, when we test the search functionality, let's make sure that we're actually getting the expected result—and only it (e.g., if we know we should get two templates matching the search string, make sure we have these two templates found and not more). Similarly, let's check that sorting does actually produce sorted results.
- Sorting perhaps calls for further optimization of our fixtures. I think it might be best to not use real data at all, but instead create by hand a list of convenient templates for the fixture that would facilitate testing. For example, we could have templates named A, B, C, D, etc., that would make sorting tests easier. We could also have templates named `String1 String2`, `String1 String3`, `String2 String3`, which would help with testing the search functionality. We could even have different fixtures for different tests—or we could create a single one, like we use now, that would be suitable for all tests.

Last but not least, there are quite a few files in `src` that are changed in this branch. Are these changes connected to the Cypress tests? If they are not, maybe we should put them in a separate pull request?

Sorry for being so slow with my reviews. I'll try to increase my pace here.

kerberizer commented:
@sergesoroka Do you think we might need a meeting to discuss the next steps?

sergesoroka commented:
that's a right idea to optimize the fixture so we can use different fixtures for different tests

sergesoroka commented:
> Are these changes connected to the Cypress tests?

yes, I've added data attribute (data-cy=element) to select html elements for testing

sergesoroka commented:
> @sergesoroka Do you think we might need a meeting to discuss the next steps?

of course, let's connect when you have time

kerberizer commented on cypress/tools/fixture-generator.py:
```diff
@@ -0,0 +1,54 @@
+  random_status = random.choice(['DRAFT', 'FINALIZED'])
+
+  template = {
+        "uri": f"https://api-test.ramanchada.ideaconsult.net/template/{random_string(21)}",
```
@sergesoroka Let's put the UUID here, just like in the real templates. Something like:

```py
uuid = str(uuid.uuid4())
template = {
        "uri": f"https://api-test.ramanchada.ideaconsult.net/template/{uuid}",
        "uuid": uuid,
```

kerberizer replied:
@sergesoroka Please fix this and regenerate `fixtures/json/bk_rcapi_templates_generated.json`.

kerberizer commented on cypress/tools/fixture-generator.py:
```diff
@@ -0,0 +1,54 @@
+def random_date(start, end):
+    delta = end - start
+    random_days = random.randint(0, delta.days)
+    return start + timedelta(days=random_days)
```
@sergesoroka We're currently only testing with up to days precision. Let's tighten the random time interval to one year, e.g.:
```py
start_date = datetime(2023, 7, 1)
end_date = datetime(2024, 7, 1)
```
And let's add random microseconds with something like:
```py
random_days = random.randint(0, delta.days)
random_us = random.randint(1, 86400000000)  # intentionally have 1 as lower bound
return start + timedelta(days=random_days, microseconds=random_us)
```

kerberizer commented on cypress/tools/fixture-generator.py:
```diff
@@ -0,0 +1,54 @@
+    return ''.join(random.choices(string.printable + 
+                                  string.digits + 
+                                  string.punctuation + 
+                                  string.whitespace, k=length))
```
@sergesoroka `string.printable` should be enough, because [it includes the other ones](https://docs.python.org/3/library/string.html#string.printable).

kerberizer commented on cypress/tools/fixture-generator.py:
```diff
@@ -0,0 +1,54 @@
+
+data = []
+
+for _ in range(15):
```
Let's produce much larger number of templates. This will allow for greater variety, which might help catch more possible edge cases. And we might want to be sure that our app is working fine even with a large number of templates on the backend. How about 100-150?

kerberizer commented on cypress/tools/fixture-generator.py:
```diff
@@ -0,0 +1,54 @@
+        "uuid": str(uuid.uuid4()),
+        "METHOD": "CFE",
+        "timestamp": date.isoformat(),
+        "PROTOCOL_CATEGORY_CODE": "TO_GENETIC_IN_VITRO_SECTION",
```
Let's find a way to randomize this properly as well.

kerberizer reviewed (changes requested):
@sergesoroka Thanks again! I really like the script. This isn't a full review, but we have been doing it incrementally so far anyway.

sergesoroka commented:
@kerberizer I've updated python script. Thanks for your comments! They are very helpful. 
I'm not sure about `random_underscores_string` function... if we really need it... but I tried to generate string with underscores like we have in real data e.g. ''NPO_1339_SECTION".

kerberizer commented on cypress/e2e/tableSorting.cy.js:
```diff
@@ -0,0 +1,51 @@
+const testURLRoot = "http://127.0.0.1:50722/templates/";
+
+const fixture = "json/templates_generated.json";
+// const fixture = "json/tbk_rcapi_template_default.json";
```
Let's clean up such commented code.

kerberizer commented on cypress/fixtures/json/templates_generated.json:
Please rename this to `bk_rcapi_templates_generated.json` and regenerate it after the generation script is updated.

kerberizer commented on cypress/e2e/copyTemplateLink.cy.js:
```diff
@@ -0,0 +1,33 @@
+const testURLRoot = "http://127.0.0.1:50722/templates/";
+const fixture = "json/templates_generated.json";
```
Please change this to `json/bk_rcapi_templates_generated.json`.

kerberizer commented on cypress/e2e/general.cy.js:
```diff
@@ -0,0 +1,97 @@
+      hostname: "api-test.ramanchada.ideaconsult.net",
+    },
+    {
+      fixture: "json/bk_rcapi_template_default.json",
```
Please change this to `json/bk_rcapi_templates_generated.json`.

kerberizer commented on cypress/e2e/generateExcel.cy.js:
```diff
@@ -0,0 +1,53 @@
+const testURLRoot = "http://127.0.0.1:50722/templates/";
+const fixture = "json/templates_generated.json";
```
Please change this to `json/bk_rcapi_templates_generated.json`.

kerberizer commented on cypress/e2e/generateExcel.cy.js:
```diff
@@ -0,0 +1,53 @@
+        hostname: "api-test.ramanchada.ideaconsult.net",
+      },
+      {
+        fixture,
```
For consistency, please either define the fixture directly here, or define the fixture of the next intercept via a constant as well.

kerberizer commented on cypress/e2e/generateExcel.cy.js:
```diff
@@ -0,0 +1,53 @@
+        hostname: "enanomapper.adma.ai",
+      },
+      {
+        fixture: "json/bk_rcapi_projects_default.json",
```
See the comment on the fixture above.

kerberizer commented on cypress/e2e/tableSorting.cy.js:
```diff
@@ -0,0 +1,51 @@
+const testURLRoot = "http://127.0.0.1:50722/templates/";
+
+const fixture = "json/templates_generated.json";
```
Please change this to `json/bk_rcapi_templates_generated.json`.

kerberizer commented on cypress/fixtures/json/bk_rcapi_template_default.json:
Please remove this file (unless there's a good reason to keep it).

kerberizer commented on cypress/fixtures/templates_generated.json:
Remove this file. It was apparently created and committed by mistake.

kerberizer commented on cypress/tools/fixture-generator.py:
```diff
@@ -0,0 +1,59 @@
+
+current_directory = os.getcwd()
+
+file_name = "templates_generated.json"
```
Please change this to `bk_rcapi_templates_generated.json`.

kerberizer commented on cypress/tools/fixture-generator.py:
```diff
@@ -0,0 +1,59 @@
+
+file_name = "templates_generated.json"
+
+parent_directory = os.path.dirname(current_directory) + "/fixtures"
```
I believe this should be `/fixtures/json`, maybe even `/cypress/fixtures/json` if we are going to run the script from the root project directory. If you are not sure, change it to `/cypress/fixtures/json` and run the script from the root directory of the project.

kerberizer commented on cypress/fixtures/json/bk_rcapi_projects_default.json:
Please rename this file to `bk_11ty_projects_default.json`. It is generated by the 11ty site, not the rcapi backend.

kerberizer commented on src/DataTable/TemplateTable.tsx:
```diff
@@ -96,6 +97,8 @@ export default function TemplateTable({ data }) {
     setIdShosen(data && rowSelection ? data[rowSelection]?.uuid : null);
   }, [rowSelection, pagination]);
 
+  // console.log(table.getRowModel());
+
```
Let's clean up such code.

kerberizer commented on cypress/downloads/396fc51e-dd55-451f-a6dd-913177abb667.json:
Is this file a result of the tests? If it is, we must be careful what we commit. In this case please remove it and add `cypress/downloads` above `cypress/screenshots` in `.gitignore`.

kerberizer commented on cypress/e2e/general.cy.js:
```diff
@@ -0,0 +1,97 @@
+      fixture: "json/bk_rcapi_template_default.json",
+    }
+  ).as("getAllTemplates");
+  cy.intercept(
+    {
+      method: "POST",
+      url: "/template",
+      hostname: "api-test.ramanchada.ideaconsult.net",
+    },
+    {
+      fixture: "json/bk_rcapi_template_default.json",
+    }
+  ).as("createNewTemplate");
```
Please scrap this intercept. It will require much more work.

kerberizer commented on cypress/e2e/general.cy.js:
```diff
@@ -0,0 +1,97 @@
+      .and("be.visible");
+  });
+
+  it("can open an existing draft for editing", () => {
+    cy.get('[data-cy="draft"]').click();
+    cy.get(".nonSelected td").eq(1).click();
+    cy.get('[data-cy="Edit blueprint"]').click();
+
+    cy.get("#Save").click();
+  });
```
Please scrap this test. It will require much more work. In it's current state it is practically pointless.

kerberizer reviewed (changes requested):
@sergesoroka Alright, let's finalize this. Please do:
1. Rebase the branch on `main` so that any already merged in `main` changes are reflected in the branch here. If you need some help, [this](https://www.freecodecamp.org/news/how-to-use-git-rebase/) might be a good start. Don't also hesitate to ask on Discord.
2. Make the requested changes to the code of the branch.

Once we're ready with this, we'll make a (hopefully) final review to have it approved for merging.

kerberizer reviewed (approved):
@sergesoroka I think we're ready with this PR. If you wish, take a look at the latest changes I've made. Most importantly, perhaps, two tests were problematic:
- Sorting wasn't working as expected. The app sorted columns alphanumerically instead of as plain text. E.g., `10abc` was sorted **after** `9abc`, because `9 < 10`. I've changed the sort function in `src/DataTable/TemplateTable.tsx` and it works as expected now: `10abc` comes before `9abc`, because `1` precedes `9` as a symbol.
- The copy to clipboard test **never** failed, because Cypress couldn't catch the assertion fail inside the async block. I've changed the test according to similar reports on the internet and it fails correctly when expected now.

If you have any questions, comments, suggestions, etc., please don't hesitate to share them. If not, let's merge the PR.

sergesoroka commented:
> @sergesoroka I think we're ready with this PR. If you wish, take a look at the latest changes I've made. Most importantly, perhaps, two tests were problematic:
> 
> * Sorting wasn't working as expected. The app sorted columns alphanumerically instead of as plain text. E.g., `10abc` was sorted **after** `9abc`, because `9 < 10`. I've changed the sort function in `src/DataTable/TemplateTable.tsx` and it works as expected now: `10abc` comes before `9abc`, because `1` precedes `9` as a symbol.
> * The copy to clipboard test **never** failed, because Cypress couldn't catch the assertion fail inside the async block. I've changed the test according to similar reports on the internet and it fails correctly when expected now.
> 
> If you have any questions, comments, suggestions, etc., please don't hesitate to share them. If not, let's merge the PR.

Thank you so much. I'll explore your changes to understand what I've missed. It will be very helpful for me.

kerberizer commented:
> Thank you so much. I'll explore your changes to understand what I've missed. It will be very helpful for me.

Thank **you** for working so hard on this!

FWIW, some of the changes are a matter of personal preference, like putting most tests in `general.cy.js` (my rationale here is to reduce duplicated code like common intercepts). Others might indeed be helpful, like [pathlib](https://docs.python.org/3/library/pathlib.html) in the fixture generator; if you haven't worked with it—it's a very convenient Python library.

Once you're satisfied with everything, please ping me so that we can finally merge it.

kerberizer commented:
@sergesoroka It seems like you've edited my comment. :)

sergesoroka commented:
> @sergesoroka It seems like you've edited my comment. :)

Sorry, by mistake :)

kerberizer commented:
Haha, no worries, I didn't even know that was possible. :D

Anyway, thank you very much again for working on this! It's a really important milestone!

sergesoroka commented:
> Haha, no worries, I didn't even know that was possible. :D
> 
> Anyway, thank you very much again for working on this! It's a really important milestone!

Yes, indeed! Thank you!
